---
title: "MODUL FMI survey"
author: "Alex Abbas"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(comment = NA, message = F, warning = F)
library(tidyverse)
library(readxl)
library(ComplexHeatmap)
library(DT)
library(knitr)
library(ggvenn)
library(UpSetR)
library(gridExtra)
library(survival)
library(survminer)
#library(ggpubr)

#source("/Volumes/GoogleDrive/My Drive/R/aaStuff.R")
#truColors = read.csv("/Volumes/GoogleDrive/My Drive/R/trubetskoyColors.csv",stringsAsFactors = F)

theme_set(theme_bw())

#  data_dir="/gne/data/obdroot/avastin/mo29112_modul/ngs/results/ctdna/data"
ctDnaDir <- "../data/ctDNA FMI/"
tissueDir <- "../data/FMI/"
Aug30fmiDir <- "/Volumes/OBD/avastin/mo29112_modul/ngs/rawdata/fmi_obd_mo29112_ngs_dna_targdna_foundationoneliquidcdx_20-2047_std_20210830/"

patsToExclude = c("2804","4850","7842","3207")

# patientData
patientData <- read_excel("../data/MO29112_Ch1 clinical-biomarker dashboard (2).xlsx")
patientData <- patientData %>%
  dplyr::rename(Arm = ARM, Response = BOR, SUBJECT.ID = "Subject ID") %>%
  select(SUBJECT.ID, Arm, Response) %>% 
  mutate(SUBJECT.ID = as.character(SUBJECT.ID))
patientData = patientData %>% filter( !(SUBJECT.ID %in% patsToExclude) )

clinicalData <- read_excel("../data/Clinical data cohort1.xlsx") %>% 
  dplyr::rename(
    SUBJECT.ID = SUBJID
  )
clinicalData = clinicalData %>% filter( !(SUBJECT.ID %in% patsToExclude) )

# ctDNA data
oapl_ctdna_File <- paste0(
  ifelse ( F & dir.exists(Aug30fmiDir), Aug30fmiDir, ctDnaDir ),
  "mo29112_ngs_dna_targdna_foundationoneliquidcdx_20-2047_n214_one-alteration-per-line-plus_20211214.txt"
)
liqOapl <- read.delim(oapl_ctdna_File,sep="\t")
ospl_ctdna_file = paste0(ctDnaDir,"mo29112_F1Lcdx_n214_ospl_aa_20211214.xlsx")
liqOspl <- read_excel(ospl_ctdna_file) %>%
  dplyr::rename(FMI.SAMPLE.ID = "FMI SAMPLE ID", SAMPLE.ID = "SAMPLE ID", SUBJECT.ID = "SUBJECT ID") %>% 
  mutate(
    SAMPLE.ID = as.character(SAMPLE.ID),
    SUBJECT.ID = as.character(SUBJECT.ID),
    SUBJECT.ID = gsub("\\.0","",SUBJECT.ID),
    Visit = gsub("\\W","",Visit),
    Cycle = gsub("\\W","",Cycle)
  )

liqOapl <- liqOapl %>%
  mutate(SAMPLE.ID = as.character(SAMPLE.ID),SUBJECT.ID = as.character(SUBJECT.ID)) %>% 
  left_join(liqOspl %>% select(FMI.SAMPLE.ID, Cycle, Visit))
# Tissue data
oapl_tissue_file <- paste0(
  ifelse ( F & dir.exists(Aug30fmiDir), Aug30fmiDir, tissueDir ),
  "mo29112_ngs_dna_targdna_foundationone_20-2041_n65_one-alteration-per-line_20200615.txt"
)
tissueOapl = read.delim(oapl_tissue_file,sep="\t") %>% 
  mutate(SUBJECT.ID = as.character(SUBJECT.ID))
ospl_tissue_file <- paste0(tissueDir,"mo29112_ngs_dna_targdna_foundationone_20-2041_n65_one-sample-per-line_20200615.xlsx")
tissueOspl <- read_excel(ospl_tissue_file) %>%
  dplyr::rename(SAMPLE.ID = "SAMPLE ID", SUBJECT.ID = "SUBJECT ID") %>% 
  mutate(SUBJECT.ID = as.character(SUBJECT.ID))

# Stack tissue and liquid data
oapl <- bind_rows(
  tissueOapl %>% mutate(Visit = "Tissue"),
  liqOapl %>% mutate(SAMPLE.ID=as.character(SAMPLE.ID))
) %>%
  mutate(
    variant = gsub("-","",paste0(SV.PROTEIN.CHANGE,CNA.TYPE,REARR.DESCRIPTION)),
    baseEnd = ifelse(grepl("C1",Visit),"C1D1",
                     ifelse(grepl("PD",Visit),"PD",NA)),
    SUBJECT.ID = as.character(SUBJECT.ID)
  ) %>%
  left_join(
    patientData
  )
oapl = oapl %>% filter( !(SUBJECT.ID %in% patsToExclude) )

ospl = bind_rows(
  liqOspl %>% mutate( SAMPLE.ID = as.character(SAMPLE.ID) ),
  tissueOspl %>% mutate( Visit = "Tissue", SUBJECT.ID = as.character(SUBJECT.ID) )
)
ospl = patientData %>% mutate( SUBJECT.ID = as.character(SUBJECT.ID) )%>%
  left_join(ospl) %>% 
  mutate(SUBJECT.ID = gsub("\\.0","",SUBJECT.ID))
ospl = ospl %>% filter( !(SUBJECT.ID %in% patsToExclude) & !(is.na(SAMPLE.ID)))

# split MSI and TMB data from variants
msiScore <- oapl %>% filter(VARIANT.TYPE=="BIOMARKER-MSI SCORE")
msiStatus <- oapl %>% filter(VARIANT.TYPE=="BIOMARKER-MSI STATUS")
tmb <- oapl %>% filter(VARIANT.TYPE=="BIOMARKER-TMB")
oapl <- oapl %>% filter(!grepl("BIOMARKER",VARIANT.TYPE))

# re-subset tissue and liq
tissueOapl <- oapl %>% filter(Visit == "Tissue")
tissueOaplAll <- tissueOapl
tissueOapl <- tissueOapl %>% filter(SOMATIC.STATUS.FUNCTIONAL.IMPACT != "unknown")

liqOapl <- oapl %>% filter(Visit != "Tissue")
liqOaplAll <- liqOapl
liqOapl <- liqOapl %>% filter(SOMATIC.STATUS.FUNCTIONAL.IMPACT != "unknown")

# remove unknown impact variants now that tissueOaplAll and liqOaplAll are made
oapl <- oapl %>% filter(SOMATIC.STATUS.FUNCTIONAL.IMPACT != "unknown")

# calculate event times in weeks
cycleToTime = function (Cycle) {
#  Cycle[Cycle == "C1D1IP"] = -16
  Cycle[grepl("IP",Cycle)] = ( as.numeric(gsub("C(\\d{1,2})D1IP","\\1",Cycle[grepl("IP",Cycle)])) - 1) * 2 - 16
  Cycle[Cycle == "Tissue"] = -43.33333
  Cycle[grepl("MP",Cycle)] = ( as.numeric(gsub("C(\\d{1,2})D1MP","\\1",Cycle[grepl("MP",Cycle)])) - 1) * 2
  return(as.numeric(Cycle))
}
oapl[oapl$Visit=="Tissue","Cycle"] = "Tissue"
oapl <- oapl %>% mutate( timepoint = cycleToTime(Cycle) )
ospl[ospl$Visit=="Tissue","Cycle"] = "Tissue"
ospl <- ospl %>% mutate( timepoint = cycleToTime(Cycle) )

pathwayGenes <- c("EGFR", "KRAS", "NRAS", "BRAF", "MAP2K1", "NF1")

```

# Introduction

MODUL (MO29112, NCT02291289) is an adaptable, phase 2, signal-seeking trial testing novel agents as first-line therapy for predefined subgroups of patients with metastatic colorectal cancer. It is comprised of four distinct cohorts. This analysis focuses on cohort 1, comprised of subjects with BRAF V600E mutant tumors, randomized to either the experimental arm of 5-FU/LV + cetuximab + vemurafenib or a control arm of FP + bevacizumab. Subjects' tumor tissue was tested with the FMI Foundation One CDx assay, and plasma samples from C1D1 and/or PD/EOT were tested with the FMI Foundation One Liquid CDx assay. Variants detected at these three points will be compared/contrasted to test hypotheses that treatment affected tumor genotype, specifically that cetuximab put a negative selective pressure on V600E mutations and a positive selective pressure on compensatory mutations elsewhere in the EGFR pathway. 

# Data

## Foundation One CDx (i.e. Tissue Assay)

Survey June 15 2020 F1CDx tissue data, from OBD-FS data file:

    mo29112_ngs_dna_targdna_foundationone_20-2041_n65_one-alteration-per-line_20200615

```{r}


genes <- names(sort(table(tissueOapl$Gene),decreasing = T))
#allgenes <- unique(oapl$Gene)
subjs <- unique(tissueOapl$SUBJECT.ID)
mat <- matrix("",nrow=length(topGenes),ncol=length(subjs))
dimnames(mat) <- list(topGenes,subjs)
topOapl <- tissueOapl %>% filter(GENE %in% topGenes)
x <- apply(topOapl,1,function(oa){
  mat[oa["GENE"],oa["SUBJECT.ID"]] <<- oa["VARIANT.TYPE"]
})
mat[grepl("\\w",mat)] = "Mut"
mat[mat==""] = "WT"
matt = data.frame( SUBJECT.ID = colnames(mat), t(mat) )
clinPlus = clinicalData %>% group_by(SUBJECT.ID, ARM) %>% summarise(ADY = max(ADY)) %>% inner_join(matt)
clinPlus = clinPlus %>% mutate(
  event = 1,
  tte = as.numeric(ADY) / 365 * 12
) %>% filter(!is.na(tte))

myKM <- function(dat, tte="tte", cen="event", indvar = "", trt = "", strataCols = c("blue","red"), ylab = "PFS probability", lty = "solid", main = "") {
  iformula <- as.formula(sprintf("Surv(%s, %s) ~ %s", tte, cen, indvar))
  resCox <- coxph(iformula, data = dat)
  hr <- signif(1/summary(resCox)$conf.int,digits=2)
  p <- summary(resCox)$logtest["pvalue"]
  coxStatText <- paste0("P = ",signif(p,digits=2), " HR = ",hr[1]," (",hr[4],"-",hr[3],")")
  fit <- surv_fit(iformula, data = dat)
  print(
    ggsurvplot(
      fit,
      pval = coxStatText,#, conf.int = TRUE,
      risk.table = TRUE, #risk.table.col = strataCols, # Add risk table
      risk.table.y.text.col = T, risk.table.y.text = FALSE,
      # linetype = lty, # Change line type by groups
      # # surv.median.line = "hv", # Specify median survival
      # # ggtheme = theme_bw(), # Change ggplot2 theme
      palette = strataCols,
      title = main
    )
  )
}


myKM(clinPlus %>% filter(ARM == "Experimental arm"),indvar = "TP53")
myKM(clinPlus %>% filter(ARM == "Control arm"),indvar = "TP53")

myKM(clinPlus %>% filter(ARM == "Experimental arm"),indvar = "SMAD4")
myKM(clinPlus %>% filter(ARM == "Control arm"),indvar = "SMAD4")

myKM(clinPlus %>% filter(ARM == "Experimental arm"),indvar = "RNF43")
myKM(clinPlus %>% filter(ARM == "Control arm"),indvar = "RNF43")

```

